File: packages/client-generator/src/runtime/paginate.ts
import type { OperationArgs } from './create-client.js';
import type { PaginationSpec, QueryValue, RequestOptions } from './types.js';

/**
 * Auto-pagination (capability module — wired into `createClient`, dispatched by the
 * method's `.pages()`/`.items()`): walk an operation's pages by advancing the descriptor's
 * `param` query parameter, per its `style`. The caller's args are never mutated — each
 * request gets a fresh `params` clone — and `init` is forwarded to every call.
 *
 * Iteration is error-mode-agnostic: `call` always resolves to the RAW page (on a
 * result-mode client the attachment unwraps the envelope first), so a failed page
 * aborts iteration by throwing `ApiError`, even on result-mode clients; the `onError`
 * middleware hook (throw-mode-only) is not invoked.
 */

/**
 * Resolve an RFC 6901 JSON pointer (`~1` → `/`, `~0` → `~`) against a value.
 * The empty pointer is the whole document; anything else must start with `/`.
 * Returns `undefined` on any miss (bad token, absent key, non-object step) — never throws.
 */
export function resolvePointer(value: unknown, pointer: string): unknown {
  if (pointer === '') return value;
  if (!pointer.startsWith('/')) return undefined;
  let current = value;
  for (const token of pointer.slice(1).split('/')) {
    const key = token.replaceAll('~1', '/').replaceAll('~0', '~');
    if (Array.isArray(current)) {
      if (!/^(0|[1-9]\d*)$/.test(key)) return undefined;
      current = current[Number(key)];
    } else if (Object(current) === current && key in (current as object)) {
      current = (current as Record<string, unknown>)[key];
    } else {
      return undefined;
    }
  }
  return current;
}

/**
 * Iterate an operation's full page results. Every page is yielded before the stop
 * condition is evaluated, so the last page always arrives. Cursor style resumes from a
 * caller-provided `params[spec.param]`, stops when the optional `hasMore` pointer
 * resolves to `false` or when `nextCursor` resolves to `undefined`/`null`/`''`, and
 * throws if the next cursor is not a string or number, or
 * if the same cursor comes back twice in a row (infinite-loop guards). Offset/page
 * styles advance by item count / by one and stop when
 * the `items` pointer misses or the array is empty.
 */
export async function* pages<TPage>(
  call: (args?: OperationArgs, init?: RequestOptions) => Promise<TPage>,
  spec: PaginationSpec,
  args: OperationArgs = {},
  init?: RequestOptions
): AsyncGenerator<TPage> {
  if (spec.style === 'cursor') {
    let cursor: unknown = args.params?.[spec.param];
    while (true) {
      const params = { ...args.params };
      if (cursor !== undefined) params[spec.param] = cursor as QueryValue;
      const page = await call({ ...args, params }, init);
      yield page;
      // Connection-style APIs keep a non-null cursor on the last page and signal the
      // end via a boolean flag — honor it before the cursor check to skip the
      // follow-up empty request. Strictly `false`: a missing pointer falls through.
      if (spec.hasMore !== undefined && resolvePointer(page, spec.hasMore) === false) return;
      const next = resolvePointer(page, spec.nextCursor);
      if (next === undefined || next === null || next === '') return;
      if (typeof next !== 'string' && typeof next !== 'number') {
        // A fresh non-scalar cursor never compares equal, so without this guard a lying
        // server would slip past the did-not-advance check into an infinite loop.
        throw new Error(`Pagination cursor at ${spec.nextCursor} is not a string or number`);
      }
      if (next === cursor) {
        throw new Error('Pagination did not advance: operation returned the same cursor twice');
      }
      cursor = next;
    }
  } else if (spec.style === 'link') {
    // `link` iteration needs the response's `Link` header, which the parsed-page `call`
    // cannot carry — the client wires those operations to `pagesByLink` instead.
    throw new Error('link-style pagination iterates via pagesByLink');
  } else {
    // Coerce the starting position to a number: a caller may pass `params[spec.param]` as a
    // string (common from URL/form input), and `+=` on a string would concatenate.
    const start = args.params?.[spec.param];
    const fallback = spec.style === 'page' ? 1 : 0;
    let position = start === undefined || Number.isNaN(Number(start)) ? fallback : Number(start);
    while (true) {
      const page = await call(
        { ...args, params: { ...args.params, [spec.param]: position } },
        init
      );
      yield page;
      const pageItems = resolvePointer(page, spec.items);
      if (!Array.isArray(pageItems) || pageItems.length === 0) return;
      position += spec.style === 'page' ? 1 : pageItems.length;
    }
  }
}

/**
 * Iterate the operation's individual items: each page's `items` pointer, flattened.
 * A cursor-style page whose pointer misses yields nothing but pagination continues;
 * for offset/page styles a miss has already stopped `pages`.
 */
export async function* items<TItem>(
  call: (args?: OperationArgs, init?: RequestOptions) => Promise<unknown>,
  spec: PaginationSpec,
  args?: OperationArgs,
  init?: RequestOptions
): AsyncGenerator<TItem> {
  for await (const page of pages(call, spec, args, init)) {
    const pageItems = resolvePointer(page, spec.items);
    if (Array.isArray(pageItems)) yield* pageItems as TItem[];
  }
}

/**
 * The per-page call the `link`-style iterators drive: the parsed page plus the raw
 * `Link` header and the page's own URL (for resolving a relative `next` target).
 */
export type LinkPageCall = (
  args?: OperationArgs,
  init?: RequestOptions
) => Promise<{ page: unknown; linkHeader: string | null; url: string }>;

/** The `rel="next"` target of an RFC 8288 `Link` header, or `undefined` when absent. */
export function linkNext(header: string | null): string | undefined {
  if (header === null) return undefined;
  // Entries are `<url>; rel="next"`, comma-separated; the URL is inside `<>`, so split
  // on commas that precede a `<` and a comma inside a target cannot break an entry.
  for (const entry of header.split(/,\s*(?=<)/)) {
    const target = /^\s*<([^>]*)>(.*)$/.exec(entry);
    if (!target) continue;
    const rel = /;\s*rel\s*=\s*"?([^";]+)"?/i.exec(target[2]);
    // `rel` may carry several space-separated relation types (RFC 8288 §3.3).
    if (rel && rel[1].split(/\s+/).includes('next')) return target[1];
  }
  return undefined;
}

/**
 * Iterate a `link`-style operation's pages: follow the `Link` header's `rel="next"`
 * target by merging ITS query params into the next call — every page goes through the
 * same declared endpoint, so auth, middleware, and `serverUrl` handling apply
 * unchanged, and credentials can never be handed to a cross-origin `next` URL.
 * Stops when no `rel="next"` is present; throws when the link does not advance
 * (the same target twice, or a self-link — an infinite-loop guard).
 */
export async function* pagesByLink<TPage>(
  call: LinkPageCall,
  args: OperationArgs = {},
  init?: RequestOptions
): AsyncGenerator<TPage> {
  let params = args.params;
  let previous: string | undefined;
  while (true) {
    const { page, linkHeader, url } = await call({ ...args, params }, init);
    yield page as TPage;
    const target = linkNext(linkHeader);
    if (target === undefined) return;
    // A relative target resolves against the page's own URL (RFC 8288 §3.1).
    const next = new URL(target, url).toString();
    if (next === previous || next === url) {
      throw new Error('Pagination did not advance: the Link rel="next" target repeats');
    }
    previous = next;
    const linkParams: Record<string, string> = {};
    for (const [key, value] of new URL(next).searchParams) linkParams[key] = value;
    params = { ...args.params, ...linkParams };
  }
}

/** Iterate a `link`-style operation's individual items: each page's `items` pointer, flattened. */
export async function* itemsByLink<TItem>(
  call: LinkPageCall,
  spec: PaginationSpec,
  args?: OperationArgs,
  init?: RequestOptions
): AsyncGenerator<TItem> {
  for await (const page of pagesByLink(call, args, init)) {
    const pageItems = resolvePointer(page, spec.items);
    if (Array.isArray(pageItems)) yield* pageItems as TItem[];
  }
}

File: packages/client-generator/src/generators/python/runtime/_paginate.py
# Auto-pagination iterators for generated Python clients — the TypeScript
# runtime's paginate.ts semantics ported: cursor (next-cursor pointer, optional
# has-more flag, repeated-cursor guard), offset/page (advance by count/one,
# repeated-page guard, null start treated as absent), and link (RFC 8288
# `Link: rel="next"` following with relative resolution and a loop guard).
from __future__ import annotations

import re
from typing import Any, AsyncIterator, Awaitable, Callable, Dict, Iterator, Optional, Tuple
from urllib.parse import parse_qsl, urljoin, urlparse

# call(params) -> (parsed_json, httpx.Response)
PageCall = Callable[[Dict[str, Any]], Tuple[Any, Any]]


def resolve_pointer(data: Any, pointer: str) -> Any:
    """RFC 6901 JSON pointer over parsed JSON; None on any miss."""
    if pointer == "":
        return data
    if not pointer.startswith("/"):
        return None
    current = data
    for token in pointer[1:].split("/"):
        key = token.replace("~1", "/").replace("~0", "~")
        if isinstance(current, dict):
            current = current.get(key)
        elif isinstance(current, list) and key.isdigit():
            index = int(key)
            current = current[index] if index < len(current) else None
        else:
            return None
        if current is None:
            return None
    return current


def iter_pages(call: PageCall, spec: Dict[str, Any], params: Optional[Dict[str, Any]] = None) -> Iterator[Any]:
    """Yield raw page JSON per the pagination spec; every page is yielded before
    the stop condition is evaluated, so the last page always arrives."""
    style = spec["style"]
    base = dict(params or {})
    if style == "cursor":
        cursor = base.get(spec["param"])
        while True:
            page_params = dict(base)
            if cursor is not None:
                page_params[spec["param"]] = cursor
            page, _response = call(page_params)
            yield page
            if spec.get("has_more") is not None and resolve_pointer(page, spec["has_more"]) is False:
                return
            nxt = resolve_pointer(page, spec.get("next_cursor", ""))
            if nxt is None or nxt == "":
                return
            if not isinstance(nxt, (str, int, float)):
                raise ValueError(f"Pagination cursor at {spec['next_cursor']} is not a string or number")
            if nxt == cursor:
                raise ValueError("Pagination did not advance: the operation returned the same cursor twice")
            cursor = nxt
    elif style == "link":
        yield from _iter_pages_by_link(call, base)
    else:  # offset / page
        start = base.get(spec["param"])
        fallback = 1 if style == "page" else 0
        try:
            position = fallback if start in (None, "") else int(start)
        except (TypeError, ValueError):
            position = fallback
        previous_items = None
        while True:
            page, _response = call({**base, spec["param"]: position})
            items = resolve_pointer(page, spec.get("items", ""))
            serialized = repr(items) if isinstance(items, list) else None
            if serialized is not None and serialized == previous_items:
                raise ValueError("Pagination did not advance: the operation returned the same page twice")
            yield page
            if not isinstance(items, list) or len(items) == 0:
                return
            previous_items = serialized
            position += 1 if style == "page" else len(items)


def _link_next(header: Optional[str]) -> Optional[str]:
    if not header:
        return None
    for entry in re.split(r",\s*(?=<)", header):
        match = re.match(r"^\s*<([^>]*)>(.*)$", entry)
        if not match:
            continue
        rel = re.search(r';\s*rel\s*=\s*"?([^";]+)"?', match.group(2), re.IGNORECASE)
        if rel and "next" in rel.group(1).split():
            return match.group(1)
    return None


def _iter_pages_by_link(call: PageCall, base: Dict[str, Any]) -> Iterator[Any]:
    params = dict(base)
    previous = None
    while True:
        page, response = call(params)
        yield page
        target = _link_next(response.headers.get("link"))
        if target is None:
            return
        page_url = str(response.request.url) if response.request is not None else ""
        nxt = urljoin(page_url or "http://relative.invalid", target)
        if nxt in (previous, page_url):
            raise ValueError('Pagination did not advance: the Link rel="next" target repeats')
        previous = nxt
        link_params: Dict[str, Any] = {}
        for key, value in parse_qsl(urlparse(nxt).query):
            if key in link_params:
                existing = link_params[key]
                link_params[key] = [*existing, value] if isinstance(existing, list) else [existing, value]
            else:
                link_params[key] = value
        params = {**base, **link_params}


def iter_items(call: PageCall, spec: Dict[str, Any], params: Optional[Dict[str, Any]] = None) -> Iterator[Any]:
    """Each page's `items` pointer, flattened."""
    for page in iter_pages(call, spec, params):
        items = resolve_pointer(page, spec.get("items", ""))
        if isinstance(items, list):
            yield from items


# call(params) -> awaitable of (parsed_json, httpx.Response)
AsyncPageCall = Callable[[Dict[str, Any]], Awaitable[Tuple[Any, Any]]]


async def aiter_pages(
    call: AsyncPageCall, spec: Dict[str, Any], params: Optional[Dict[str, Any]] = None
) -> AsyncIterator[Any]:
    """Async mirror of iter_pages — same stop conditions and guards."""
    style = spec["style"]
    base = dict(params or {})
    if style == "cursor":
        cursor = base.get(spec["param"])
        while True:
            page_params = dict(base)
            if cursor is not None:
                page_params[spec["param"]] = cursor
            page, _response = await call(page_params)
            yield page
            if spec.get("has_more") is not None and resolve_pointer(page, spec["has_more"]) is False:
                return
            nxt = resolve_pointer(page, spec.get("next_cursor", ""))
            if nxt is None or nxt == "":
                return
            if not isinstance(nxt, (str, int, float)):
                raise ValueError(f"Pagination cursor at {spec['next_cursor']} is not a string or number")
            if nxt == cursor:
                raise ValueError("Pagination did not advance: the operation returned the same cursor twice")
            cursor = nxt
    elif style == "link":
        previous = None
        link_params: Dict[str, Any] = dict(base)
        while True:
            page, response = await call(link_params)
            yield page
            target = _link_next(response.headers.get("link"))
            if target is None:
                return
            page_url = str(response.request.url) if response.request is not None else ""
            nxt = urljoin(page_url or "http://relative.invalid", target)
            if nxt in (previous, page_url):
                raise ValueError('Pagination did not advance: the Link rel="next" target repeats')
            previous = nxt
            merged: Dict[str, Any] = {}
            for key, value in parse_qsl(urlparse(nxt).query):
                if key in merged:
                    existing = merged[key]
                    merged[key] = [*existing, value] if isinstance(existing, list) else [existing, value]
                else:
                    merged[key] = value
            link_params = {**base, **merged}
    else:
        start = base.get(spec["param"])
        fallback = 1 if style == "page" else 0
        try:
            position = fallback if start in (None, "") else int(start)
        except (TypeError, ValueError):
            position = fallback
        previous_items = None
        while True:
            page, _response = await call({**base, spec["param"]: position})
            items = resolve_pointer(page, spec.get("items", ""))
            serialized = repr(items) if isinstance(items, list) else None
            if serialized is not None and serialized == previous_items:
                raise ValueError("Pagination did not advance: the operation returned the same page twice")
            yield page
            if not isinstance(items, list) or len(items) == 0:
                return
            previous_items = serialized
            position += 1 if style == "page" else len(items)


async def aiter_items(
    call: AsyncPageCall, spec: Dict[str, Any], params: Optional[Dict[str, Any]] = None
) -> AsyncIterator[Any]:
    async for page in aiter_pages(call, spec, params):
        items = resolve_pointer(page, spec.get("items", ""))
        if isinstance(items, list):
            for item in items:
                yield item
